hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -117,27 +117,55 @@ public FileIO io() {
   @Override
   protected void doRefresh() {
     String metadataLocation = null;
+    String previousMetadataLocation = null;
+    Table table = null;
     try {
-      Table table = metaClients.run(client -> client.getTable(database, tableName));
+      table = metaClients.run(client -> client.getTable(database, tableName));
       validateTableIsIceberg(table, fullName);
 
       metadataLocation = table.getParameters().get(METADATA_LOCATION_PROP);
-
+      previousMetadataLocation = table.getParameters().get(PREVIOUS_METADATA_LOCATION_PROP);
     } catch (NoSuchObjectException e) {
       if (currentMetadataLocation() != null) {
         throw new NoSuchTableException("No such table: %s.%s", database, tableName);
       }
-
     } catch (TException e) {
       String errMsg = String.format("Failed to get table info from metastore %s.%s", database, tableName);
       throw new RuntimeException(errMsg, e);
-
     } catch (InterruptedException e) {
       Thread.currentThread().interrupt();
       throw new RuntimeException("Interrupted during refresh", e);
     }
 
-    refreshFromMetadataLocation(metadataLocation);
+    refreshAndFixNoFoundMetadataLocation(metadataLocation, previousMetadataLocation, table);
+  }
+
+  private void refreshAndFixNoFoundMetadataLocation(String metadataLocation, String previousMetadataLocation,
+      Table table) {
+    try {
+      // if the current metadataLocation is unavaliable ,we will replace it with previousMetadataLocation in case of
+      // both metadataLocation no found.
+      if (refreshFromMetadataLocation(metadataLocation, previousMetadataLocation) && table != null) {
+        Map<String, String> parameters = table.getParameters();
+        parameters.put(METADATA_LOCATION_PROP, previousMetadataLocation);
+        persistTable(table, true);
```

Comment:
This introduces a write operation in the Iceberg read path of a table. I feel that it is not a good idea for readers to modify table metadata in the metastore. Would it be okay to not try to fix the table during reads? The readers can still probably use the previousMetadataLocation for loads. To solve the issue you see in https://github.com/apache/iceberg/issues/1688 would it better if the writer itself fixes the issue?

Reply:
> Committing this change introduces a correctness problem. I don't think this should be committed. It is not safe to automatically roll back if there is a problem loading the current metadata file because the changes in the current metadata file are lost.
> 
> There are cases where the current metadata file may not be found. The S3A file system may throw `NotFoundException` because of negative caching in S3, even though the metadata file exists. If that happens, then this would silently roll back anything committed in the current metadata. There are other cases where this may happen as well and we simply don't know why the current metadata file was deleted. Assuming a specific cause is not a viable solution.
> 
> The problem in #1688 is that the Hive table was updated, but the operation appeared to fail on the client. The client deletes the metadata file from the commit attempt so that the commit cannot be accidentally successful (the case you hit) because the client will throw an exception that will very likely result in a retry. For example, Flink appending files will keep trying to append the same files.
> 
> I think the right solution to the problem is to improve the handling in the commit path, not the read path. The commit path can load the table again and check whether the metadata file is the one it attempted to commit. If it is, then it can return success instead of deleting the metadata file.

@shardulm94  thanks for your suggestions .I'll try some improvments on writing.